crates/holochain/src/core/sys_validate.rs

```diff
@@ -391,3 +398,317 @@ pub fn check_update_reference(
         .into())
     }
 }
+
+/// If we are not holding this header then
+/// retrieve it and send it as a RegisterAddLink DhtOp
+/// to our incoming_dht_ops_workflow.
+///
+/// Apply a checks callback to the Element.
+///
+/// Additionally sys validation will be triggered to
+/// run again if we weren't holding it.
+pub async fn check_and_hold_register_add_link<F>(
+    hash: &HeaderHash,
+    workspace: &mut SysValidationWorkspace,
+    network: HolochainP2pCell,
+    incoming_dht_ops_sender: IncomingDhtOpSender,
+    f: F,
+) -> SysValidationResult<()>
+where
+    F: FnOnce(&Element) -> SysValidationResult<()>,
+{
+    let source = check_and_hold(hash, workspace, network).await?;
+    f(source.as_ref())?;
+    incoming_dht_ops_sender
+        .send_register_add_link(source)
+        .await?;
+    Ok(())
+}
```

Comment:
I would approve of a little macro to generate all these functions, but this explicitness is cool too